### DISCUSSION TITLE
Issue?  More-Derived Extension Method

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#2579] [Q&A] [answered]
I have this class:
````
    public static class CollectionExtensions {
        public static void Add<T>(this ICollection<T> This, IEnumerable<T> Items) {
            foreach (var item in Items) {
                This.Add(item);
            }
        }
    }
````

And I call it like so:
````
            var C = new ObservableCollection<object>();
            C.Add("Foo");

            var D = new List<Object> { };
//THIS LINE
            C.Add(D);
````

When I run the code, the compiler selects ObservableCollection.Add instead of CollectionExtensions.Add.  Why is that?  I know that ObservableCollection.Add is a valid match but I believe that CollectionExtensions.Add is a better match.

**Comment by HaloFour** [comment, marked as answer]
IIRC extension methods are not considered alongside instance methods in overload resolution.  The compiler will only consider extension methods if there are no instance methods that can be used given the arguments provided.

I'd suggest renaming your extension method to `AddRange`, that is the convention used to add multiple elements to a collection in the BCL.

**Comment by svick** [comment]
[Extension methods are considered only if there is no valid instance method:](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#extension-method-invocations)

> In a method invocation […] if the normal processing of the invocation finds no applicable methods, an attempt is made to process the construct as an extension method invocation.

Since `ObservableCollection.Add` is applicable, extension methods are not even considered here.

**Comment by TonyValenti** [comment]
Thanks @HaloFour !
How would I use my custom Add/AddRange using the collection initializer syntax?

**Comment by yaakov-h** [comment]
I don't think you can.

Collection initializer syntax is designed to enumerate each individual element you want to add to the collection, not a whole bunch at once:

>  Each element initializer specifies an element to be added to the collection object being initialized - [spec](https://github.com/dotnet/csharplang/blob/b6872068faa1b34ae3c9585d182908729b10aae4/spec/expressions.md#collection-initializers)

You would have to actually call `Add`/`AddRange` yourself, the compiler can't do that for you here.

**Comment by CyrusNajmabadi** [comment]
This was very much by design (and effectively mandated by extreme force by the community).  People felt it was totally unacceptable (and made that utterly and completely unambiguous when we introduced 'extensions') that an 'extension' would ever get priority over their own types' authored members.

If you don't have this, then random extensions added to some namespace you include can now change the behavior of your types and how they were directly authored.
